src/Evolution/Systems/ScalarWave/BoundaryConditions/DirichletAnalytic.hpp

```diff
@@ -0,0 +1,110 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <memory>
+#include <optional>
+#include <pup.h>
+#include <string>
+#include <type_traits>
+
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Evolution/BoundaryConditions/Type.hpp"
+#include "Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp"
+#include "Evolution/Systems/ScalarWave/Tags.hpp"
+#include "Options/Options.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "PointwiseFunctions/AnalyticData/Tags.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/AnalyticSolution.hpp"
+#include "Time/Tags.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+namespace domain::Tags {
+template <size_t Dim, typename Frame>
+struct Coordinates;
+}  // namespace domain::Tags
+/// \endcond
+
+namespace ScalarWave::BoundaryConditions {
+/*!
+ * \brief Sets Dirichlet boundary conditions using the analytic solution or
+ * analytic data.
+ */
+template <size_t Dim>
+class DirichletAnalytic final : public BoundaryCondition<Dim> {
```

Comment:
Do you think it's possible for evolution systems to define a general `DirichletAnalytic` boundary condition template that each system can choose to include? That would presumably reduce the amount of code+tests you have to write for each system by quite a lot :)

Reply:
No, at least not without making the generic code basically impossible to understand. The boundary conditions depend on the system, the boundary correction, possible auxiliary arguments like constraint damping parameters/functions, as well as the background (e.g. GRMHD). This would mean dissecting a list of arguments into a bunch of tuples, figuring out how to compute whatever is needed (e.g. getting the lapse from the spacetime metric, but how do I figure out that's how to compute it in generic code?), and then returning the result. At best, that would be probably 1500 lines of just manipulating typelists, parameter packs, and tuples.